grpc-server/tinkerbell.go

```diff
@@ -2,47 +2,219 @@ package grpcserver
 
 import (
 	"context"
+	"fmt"
+	"log"
+	"time"
 
-	exec "github.com/tinkerbell/tink/executor"
+	"github.com/tinkerbell/tink/db"
 	pb "github.com/tinkerbell/tink/protos/workflow"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
+var workflowData = make(map[string]int)
+
 // GetWorkflowContexts implements tinkerbell.GetWorkflowContexts
 func (s *server) GetWorkflowContexts(req *pb.WorkflowContextRequest, stream pb.WorkflowSvc_GetWorkflowContextsServer) error {
-	return exec.GetWorkflowContexts(req, stream, s.db)
+	wfs, err := getWorkflowsForWorker(s.db, req.WorkerId)
+	if err != nil {
+		return err
+	}
+	for _, wf := range wfs {
+		wfContext, err := s.db.GetWorkflowContexts(context.Background(), wf)
+		if err != nil {
+			return status.Errorf(codes.Aborted, "invalid workflow %s found for worker %s", wf, req.WorkerId)
+		}
+		if isApplicableToSend(context.Background(), wfContext, req.WorkerId, s.db) {
+			if err := stream.Send(wfContext); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
 }
 
 // GetWorkflowContextList implements tinkerbell.GetWorkflowContextList
 func (s *server) GetWorkflowContextList(context context.Context, req *pb.WorkflowContextRequest) (*pb.WorkflowContextList, error) {
-	return exec.GetWorkflowContextList(context, req, s.db)
+	wfs, err := getWorkflowsForWorker(s.db, req.WorkerId)
+	if err != nil {
+		return nil, err
+	}
+	wfContexts := []*pb.WorkflowContext{}
+	for _, wf := range wfs {
+		wfContext, err := s.db.GetWorkflowContexts(context, wf)
+		if err != nil {
+			return nil, status.Errorf(codes.Aborted, "Invalid workflow %s found for worker %s", wf, req.WorkerId)
+		}
+		wfContexts = append(wfContexts, wfContext)
+	}
+	return &pb.WorkflowContextList{
+		WorkflowContexts: wfContexts,
+	}, nil
 }
 
 // GetWorkflowActions implements tinkerbell.GetWorkflowActions
 func (s *server) GetWorkflowActions(context context.Context, req *pb.WorkflowActionsRequest) (*pb.WorkflowActionList, error) {
-	return exec.GetWorkflowActions(context, req, s.db)
+	wfID := req.GetWorkflowId()
```

Comment:
`WorkflowActionsRequest` struct has the `WorkflowId` field is described as omit empty in `workflow.pb.go`. If `req` does not have this field then `GetWorkflowId()` call would panic as this method does not check whether the field is empty or not. Though it does check whether req is empty or not but it does not serve the purpose. 

For you implementation to work correctly, I would suggest you to add a check in `GetWorkflowId` to verify if req has field `WorkflowId` or not.

Reply:
I believe you are referring to this one [here](https://github.com/tinkerbell/tink/blob/master/protos/workflow/workflow.pb.go#L731):
```
WorkflowId string `protobuf:"bytes,1,opt,name=workflow_id,json=workflowId,proto3" json:"workflow_id,omitempty"`
```
The `omitempty` is for JSON encoding and doesn't affect getting the value in Go code (like `req.GetWorkflowId()`). So, if workflow id is not set, the method call will always return `""` (an empty string).

Reply:
I was talking about this method. https://github.com/tinkerbell/tink/blob/master/protos/workflow/workflow.pb.go#L480 

There is no error handling to check if `WorkflowId` is not present and eventually it will panic. 

```
func (x *WorkflowContext) GetWorkflowId() string {
	if x != nil {
		return x.WorkflowId
	}
	return ""
}
```